pkg/streamline/applier/wave.go

```diff
@@ -425,6 +424,42 @@ func (in *WaveProcessor) onApply(ctx context.Context, resource unstructured.Unst
 	in.componentChan <- lo.FromPtr(common.ToComponentAttributes(appliedResource))
 }
 
+func (in *WaveProcessor) doApply(ctx context.Context, c dynamic.ResourceInterface, u unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	appliedResource, err := c.Apply(ctx, u.GetName(), &u, metav1.ApplyOptions{
+		FieldManager: smcommon.ClientFieldManager,
+		Force:        true,
+		DryRun:       lo.Ternary(in.dryRun, []string{metav1.DryRunAll}, []string{}),
+	})
+
+	// Return early if no error occurred,
+	// if the service is in dry run mode
+	// or if the resource does not have the force sync option enabled.
+	if err == nil || in.dryRun || !smcommon.HasForceSyncOption(u) {
+		return appliedResource, err
+	}
+
+	// Mark the resource as with resync in progress to use custom logic in status synchronizer.
+	patch := unstructured.Unstructured{
+		Object: map[string]any{
+			"metadata": map[string]any{
+				"annotations": map[string]any{
+					smcommon.ResyncInProgressAnnotation: "true",
+				},
+			},
+		},
+	}
+	patchBytes, _ := patch.MarshalJSON()
+	if _, err := c.Patch(ctx, u.GetName(), types.MergePatchType, patchBytes, metav1.PatchOptions{}); err != nil {
+		klog.V(log.LogLevelDebug).ErrorS(err, "failed to patch resource with resync annotation", "resource", u.GetName())
+	}
+
+	// Otherwise force sync by deleting the resource and letting it recreate in the next round.
```

Comment:
we should do this in the same round (that also means you don't need to do this unsynced book-keeping)